src/runtime-rs/src/shim/src/task_service.rs

```diff
@@ -0,0 +1,848 @@
+        let exec_id = req.exec_id.clone();
+        let s = self.sandbox.clone();
+        let mut sandbox = s.lock().unwrap();
+
+        info!(
+            "try to start container {} process {} in shimv2",
+            &cid, &exec_id
+        );
+        sandbox
+            .start_container(req.get_id(), req.get_exec_id())
+            .map_err(|e| {
+                get_rpc_status(
+                    Code::INTERNAL,
+                    format!(
+                        "Failed to Start Container {} process {} with error: {:?}",
+                        &cid, &exec_id, e
+                    ),
+                )
+            })?;
+
+        rsp.set_pid(self.pid);
+
+        info!("start container {} process {} successfully", cid, exec_id);
+        Ok(rsp)
+    }
+
+    fn do_delete(&self, req: shim::DeleteRequest) -> Result<shim::DeleteResponse> {
+        let mut rsp = shim::DeleteResponse::new();
+        let cid = req.id.clone();
+        let exec_id: String = req.exec_id;
+        let s = self.sandbox.clone();
+        let mut sandbox = s.lock().unwrap();
+
+        let ret = sandbox.remove_container(&cid, &exec_id).map_err(|e| {
+            get_rpc_status(Code::INTERNAL, format!("failed to remove container {:}", e))
+        });
+
+        if let Err(e) = ret {
+            return Err(e);
+        }
+        rsp.set_pid(self.pid);
+        Ok(rsp)
+    }
+
+    fn do_pids(&self, _req: shim::PidsRequest) -> Result<shim::PidsResponse> {
+        let mut p_info = ProcessInfo::new();
+        p_info.set_pid(self.pid);
+        let mut rsp = shim::PidsResponse::new();
+        rsp.mut_processes().push(p_info);
+
+        Ok(rsp)
+    }
+
+    fn do_pause(&self, req: shim::PauseRequest) -> Result<Empty> {
+        let cid = &req.id;
+        let mut sandbox = self.sandbox.lock().unwrap();
+
+        sandbox.pause_container(cid).map_err(|e| {
+            get_rpc_status(
+                Code::INTERNAL,
+                "Failed to Pause Container".to_string() + &e.to_string(),
+            )
+        })?;
+        Ok(Empty::new())
+    }
+
+    fn do_resume(&self, req: shim::ResumeRequest) -> Result<Empty> {
+        let cid = &req.id;
+        let mut sandbox = self.sandbox.lock().unwrap();
+
+        sandbox.resume_container(cid).map_err(|e| {
+            get_rpc_status(
+                Code::INTERNAL,
+                "Failed to resume container".to_string() + &e.to_string(),
+            )
+        })?;
+        Ok(Empty::new())
+    }
+
+    fn do_kill(&self, req: shim::KillRequest) -> Result<Empty> {
+        let cid = &req.id;
+        let all = req.all;
+        let eid = &req.exec_id;
+        let signal: Signal = Signal::try_from(req.signal as i32).unwrap();
+
+        let mut sandbox = self.sandbox.lock().unwrap();
+
+        sandbox
+            .signal_process(cid, eid, signal, all)
+            .map_err(|e| {
+                warn!("signal cid {} eid {} process err {:?}", cid, eid, e);
+            })
```

Comment:
This looks slightly odd because if `do_exec()` fails, we return an error, but if `do_kill()` fails, we return success. 
Shouldn't the caller decide whether to ignore the error or not? Returning success for most scenarios also makes unit testing almost impossible (unless you're prepared to redirect the logger and parse the log calls).

Finally, I'm assuming the signal is included in `e`?

Reply:
Fixed in the first revision.